File: go-flood/pkg/tilings/snub_square.go
package tilings

import (
	"fmt"
	"go-flood/pkg/core"
	"math"
)

func GenerateSnubSquareBoard(options Options) core.Board {
	a := options.TileSize
	// Lattice constant for snub square (3.3.4.3.4) where squares share vertices
	D := a * math.Sqrt(2.0+math.Sqrt(3.0))
	alpha := 15.0 * math.Pi / 180.0
	cols, rows := options.Cols, options.Rows
	colorCount := options.ColorCount
	rng := options.RNG

	tiles := []core.Tile{}
	squareMap := make(map[string]int)

	getSquareCenter := func(q, r int) (float64, float64) {
		return D * float64(q), D * float64(r)
	}

	idCounter := 0
	type vertexInfo struct {
		id int
		p  core.Point
	}
	vertexMap := make(map[string]vertexInfo)
	getVertexId := func(p core.Point) int {
		key := fmt.Sprintf("%.3f,%.3f", p[0], p[1])
		if info, ok := vertexMap[key]; ok {
			return info.id
		}
		id := len(vertexMap)
		vertexMap[key] = vertexInfo{id, p}
		return id
	}

	// 1. Generate Squares
	Rsq := a / math.Sqrt(2.0)
	// Track vIds for connectivity
	type vTile struct {
		id   int
		vIds []int
	}
	var vTiles []vTile

	for r := 0; r < rows; r++ {
		for q := 0; q < cols; q++ {
			cx, cy := getSquareCenter(q, r)
			id := idCounter
			idCounter++

			points := make([]core.Point, 4)
			vIds := make([]int, 4)
			for i := 0; i < 4; i++ {
				vAngle := alpha + math.Pi/4.0 + float64(i)*math.Pi/2.0
				p := core.Point{cx + Rsq*math.Cos(vAngle), cy + Rsq*math.Sin(vAngle)}
				points[i] = p
				vIds[i] = getVertexId(p)
			}

			tiles = append(tiles, core.Tile{
				ID:        id,
				ColorID:   int(rng() * float64(colorCount)),
				OwnerID:   nil,
				Points:    points,
				Neighbors: []int{},
			})
			vTiles = append(vTiles, vTile{id, vIds})
			squareMap[fmt.Sprintf("%d,%d", q, r)] = id
		}
	}

	// 2. Generate Triangles
	var allVertices []core.Point
	verticesById := make([]core.Point, len(vertexMap))
	for _, info := range vertexMap {
		verticesById[info.id] = info.p
	}
	allVertices = verticesById

	distSq := func(p1, p2 core.Point) float64 {
		dx := p1[0] - p2[0]
		dy := p1[1] - p2[1]
		return dx*dx + dy*dy
	}

	targetSq := a * a
	tolerance := 0.1 * targetSq

	type cell struct{ x, y int }
	grid := make(map[cell][]int)
	getCell := func(p core.Point) cell {
		return cell{int(math.Floor(p[0] / a)), int(math.Floor(p[1] / a))}
	}
	for i, v := range allVertices {
		c := getCell(v)
		grid[c] = append(grid[c], i)
	}

	for i := 0; i < len(allVertices); i++ {
		v1 := allVertices[i]
		c1 := getCell(v1)

		for dx := -1; dx <= 1; dx++ {
			for dy := -1; dy <= 1; dy++ {
				c2 := cell{c1.x + dx, c1.y + dy}
				for _, j := range grid[c2] {
					if j <= i {
						continue
					}
					v2 := allVertices[j]
					if math.Abs(distSq(v1, v2)-targetSq) < tolerance {
						for dx2 := -1; dx2 <= 1; dx2++ {
							for dy2 := -1; dy2 <= 1; dy2++ {
								c3 := cell{c1.x + dx2, c1.y + dy2}
								for _, k := range grid[c3] {
									if k <= j {
										continue
									}
									v3 := allVertices[k]
									if math.Abs(distSq(v1, v3)-targetSq) < tolerance && math.Abs(distSq(v2, v3)-targetSq) < tolerance {
										id := idCounter
										idCounter++
										tiles = append(tiles, core.Tile{
											ID:        id,
											ColorID:   int(rng() * float64(colorCount)),
											OwnerID:   nil,
											Points:    []core.Point{v1, v2, v3},
											Neighbors: []int{},
										})
										vTiles = append(vTiles, vTile{id, []int{i, j, k}})
									}
								}
							}
						}
					}
				}
			}
		}
	}

	// 3. Optimized Connectivity
	vertexToTiles := make([][]int, len(allVertices))
	for tileIdx, vt := range vTiles {
		for _, vId := range vt.vIds {
			vertexToTiles[vId] = append(vertexToTiles[vId], tileIdx)
		}
	}

	for i := range tiles {
		neighborCounts := make(map[int]int)
		for _, vId := range vTiles[i].vIds {
			for _, j := range vertexToTiles[vId] {
				if i == j {
					continue
				}
				neighborCounts[j]++
			}
		}
		for j, count := range neighborCounts {
			if count >= 2 {
				tiles[i].Neighbors = append(tiles[i].Neighbors, tiles[j].ID)
			}
		}
	}

	// 4. Finalize
	minX, minY := math.MaxFloat64, math.MaxFloat64
	maxX, maxY := -math.MaxFloat64, -math.MaxFloat64
	for _, t := range tiles {
		for _, p := range t.Points {
			minX = math.Min(minX, p[0])
			minY = math.Min(minY, p[1])
			maxX = math.Max(maxX, p[0])
			maxY = math.Max(maxY, p[1])
		}
	}

	for i := range tiles {
		for j := range tiles[i].Points {
			tiles[i].Points[j][0] -= minX
			tiles[i].Points[j][1] -= minY
		}
	}

	startTileIds := []int{}
	corners := []struct{ q, r int }{
		{0, 0}, {cols - 1, 0}, {0, rows - 1}, {cols - 1, rows - 1},
	}
	for _, c := range corners {
		if id, ok := squareMap[fmt.Sprintf("%d,%d", c.q, c.r)]; ok {
			startTileIds = append(startTileIds, id)
		}
	}

	return core.Board{
		Version:      1,
		Generator:    "snub-square",
		Width:        maxX - minX,
		Height:       maxY - minY,
		Cols:         cols,
		Rows:         rows,
		Tiles:        tiles,
		StartTileIds: startTileIds,
	}
}

File: js/tilings/snub-square.js
/**
 * Generates a Snub Square tiling board (3.3.4.3.4).
 * This tiling consists of squares and triangles.
 * Every vertex is shared by two squares and three triangles.
 */
export function generateSnubSquareBoard(options) {
  const { cols, rows, tileSize: a, colorCount, rng } = options;

  // Lattice constant for snub square (3.3.4.3.4)
  const D = a * Math.sqrt(2 + Math.sqrt(3));
  const alpha = 15 * Math.PI / 180;

  let tiles = [];
  const startTileIds = [];

  let idCounter = 0;
  const vertexMap = new Map();
  const vertexEdgeKey = (a, b) => a < b ? `${a},${b}` : `${b},${a}`;
  const getVertexId = (p) => {
    const key = `${p[0].toFixed(3)},${p[1].toFixed(3)}`;
    if (vertexMap.has(key)) return vertexMap.get(key).id;
    const id = vertexMap.size;
    vertexMap.set(key, { id, p });
    return id;
  };

  const Rsq = a / Math.sqrt(2);

  // 1. Generate Squares
  // There are two orientations of squares in the snub square tiling.
  const minBound = -D / 2;
  const maxXBound = (cols - 0.5) * D;
  const maxYBound = (rows - 0.5) * D;

  for (let r = -1; r <= rows; r++) {
    for (let q = -1; q <= cols; q++) {
      // Set 1: At (D*q, D*r) with rotation alpha
      const cx1 = D * q;
      const cy1 = D * r;

      // Set 2: At (D*(q+0.5), D*(r+0.5)) with rotation -alpha (or alpha + 90?)
      // Actually the snub square has all squares same rotation?
      // No, let's check: in s{4,4} squares are rotated by alpha.
      // Wait, the Wikipedia image shows all squares have same orientation relative to the lattice.
      // BUT if I only place them at (D*q, D*r), there are huge gaps.
      // The other set of squares is at (D*(q+0.5), D*(r+0.5))

      const centers = [
        { x: cx1, y: cy1, rot: alpha },
        { x: D * (q + 0.5), y: D * (r + 0.5), rot: -alpha }
      ];

      for (const center of centers) {
        const points = [];
        const vIds = [];
        for (let i = 0; i < 4; i++) {
          const vAngle = center.rot + Math.PI / 4 + i * Math.PI / 2;
          const p = [center.x + Rsq * Math.cos(vAngle), center.y + Rsq * Math.sin(vAngle)];
          points.push(p);
          vIds.push(getVertexId(p));
        }

        // Only add if within roughly [0, cols*D] x [0, rows*D]
        const centroidX = center.x;
        const centroidY = center.y;

        if (centroidX >= minBound && centroidX <= maxXBound && centroidY >= minBound && centroidY <= maxYBound) {
          const id = idCounter++;
          tiles.push({
            id,
            colorId: Math.floor(rng() * colorCount),
            ownerId: null,
            points,
            vIds,
            neighbors: []
          });
          if (q >= 0 && q < cols && r >= 0 && r < rows) {
             // Candidates for start tiles
          }
        }
      }
    }
  }

  const squareEdgeKeys = new Set();
  tiles.forEach(tile => {
    for (let i = 0; i < tile.vIds.length; i++) {
      squareEdgeKeys.add(vertexEdgeKey(tile.vIds[i], tile.vIds[(i + 1) % tile.vIds.length]));
    }
  });

  const squareEdgeCount = (vIds) => [
    vertexEdgeKey(vIds[0], vIds[1]),
    vertexEdgeKey(vIds[1], vIds[2]),
    vertexEdgeKey(vIds[2], vIds[0])
  ].filter(key => squareEdgeKeys.has(key)).length;

  // 2. Generate Triangles by finding equilateral triplets among deduplicated vertices
  const allVertices = Array.from(vertexMap.values()).map(v => v.p);
  const distSq = (p1, p2) => {
    const dx = p1[0] - p2[0];
    const dy = p1[1] - p2[1];
    return dx * dx + dy * dy;
  };

  const targetSq = a * a;
  const tolerance = 0.001 * targetSq;

  const grid = new Map();
  const getCellKey = (p) => `${Math.floor(p[0] / a)},${Math.floor(p[1] / a)}`;
  allVertices.forEach((v, idx) => {
    const key = getCellKey(v);
    if (!grid.has(key)) grid.set(key, []);
    grid.get(key).push(idx);
  });

  for (let i = 0; i < allVertices.length; i++) {
    const v1 = allVertices[i];
    const cx = Math.floor(v1[0] / a);
    const cy = Math.floor(v1[1] / a);

    for (let dx = -1; dx <= 1; dx++) {
      for (let dy = -1; dy <= 1; dy++) {
        const cell2 = grid.get(`${cx + dx},${cy + dy}`);
        if (!cell2) continue;
        for (const j of cell2) {
          if (j <= i) continue;
          const v2 = allVertices[j];
          if (Math.abs(distSq(v1, v2) - targetSq) < tolerance) {
            for (let dx2 = -1; dx2 <= 1; dx2++) {
              for (let dy2 = -1; dy2 <= 1; dy2++) {
                const cell3 = grid.get(`${cx + dx2},${cy + dy2}`);
                if (!cell3) continue;
                for (const k of cell3) {
                  if (k <= j) continue;
                  const v3 = allVertices[k];
                  if (Math.abs(distSq(v1, v3) - targetSq) < tolerance && Math.abs(distSq(v2, v3) - targetSq) < tolerance) {
                    const centroidX = (v1[0] + v2[0] + v3[0]) / 3;
                    const centroidY = (v1[1] + v2[1] + v3[1]) / 3;

                    const vIds = [i, j, k];
                    const insideHalo = centroidX >= minBound - a && centroidX <= maxXBound + a && centroidY >= minBound - a && centroidY <= maxYBound + a;

                    if (insideHalo && squareEdgeCount(vIds) >= 1) {
                      tiles.push({
                        id: idCounter++,
                        colorId: Math.floor(rng() * colorCount),
                        ownerId: null,
                        points: [v1, v2, v3],
                        vIds,
                        neighbors: []
                      });
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }

  // 3. Optimized Connectivity
  const vertexToTiles = new Map();
  tiles.forEach((tile, idx) => {
    tile.vIds.forEach(vId => {
      if (!vertexToTiles.has(vId)) vertexToTiles.set(vId, []);
      vertexToTiles.get(vId).push(idx);
    });
  });

  tiles.forEach((tile, i) => {
    const neighborCounts = new Map();
    tile.vIds.forEach(vId => {
      vertexToTiles.get(vId).forEach(j => {
        if (i === j) return;
        neighborCounts.set(j, (neighborCounts.get(j) || 0) + 1);
      });
    });
    neighborCounts.forEach((count, j) => {
      if (count >= 2) {
        tile.neighbors.push(tiles[j].id);
      }
    });
  });

  const idMap = new Map();
  const prunedTiles = tiles.filter(tile => !(tile.points.length === 3 && tile.neighbors.length <= 1));
  prunedTiles.forEach((tile, index) => idMap.set(tile.id, index));
  tiles = prunedTiles.map((tile, index) => ({
    ...tile,
    id: index,
    neighbors: tile.neighbors.filter(nId => idMap.has(nId)).map(nId => idMap.get(nId))
  }));

  // 4. Finalize
  let minX = Infinity, minY = Infinity;
  let maxX = -Infinity, maxY = -Infinity;
  for (const t of tiles) {
    for (const p of t.points) {
      minX = Math.min(minX, p[0]);
      minY = Math.min(minY, p[1]);
      maxX = Math.max(maxX, p[0]);
      maxY = Math.max(maxY, p[1]);
    }
  }

  tiles.forEach(t => {
    t.points = t.points.map(p => [p[0] - minX, p[1] - minY]);
  });

  // Pick start tiles from corners
  const findClosest = (tx, ty) => {
    let bestId = 0;
    let minDist = Infinity;
    tiles.forEach(t => {
      let cx = 0, cy = 0;
      t.points.forEach(p => { cx += p[0]; cy += p[1]; });
      cx /= t.points.length;
      cy /= t.points.length;
      const d = Math.pow(cx - tx, 2) + Math.pow(cy - ty, 2);
      if (d < minDist) {
        minDist = d;
        bestId = t.id;
      }
    });
    return bestId;
  };

  const w = maxX - minX;
  const h = maxY - minY;
  startTileIds.push(findClosest(0, 0));
  startTileIds.push(findClosest(w, 0));
  startTileIds.push(findClosest(0, h));
  startTileIds.push(findClosest(w, h));

  return {
    version: 1,
    generator: "snub-square",
    width: w,
    height: h,
    tiles,
    startTileIds
  };
}
